File: scala_learning/TestSpark/src/main/javacode/cn/wendaocp/java/spark/WordCount.java
package cn.wendaocp.java.spark;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;

import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

/**
 * @author liuqiang
 * @since 2020/4/29 12:59
 */
public class WordCount {

    public static void main(String[] args) {

    }

    public static void wc(String[] args) {
        SparkConf conf = new SparkConf();
        conf.setMaster("local");
        conf.setAppName("wc");

        JavaSparkContext sc = new JavaSparkContext(conf);
        sc.setLogLevel("WARN");

        JavaRDD<String> lines = sc.textFile("./data/words");
        JavaRDD<String> words = lines.flatMap((FlatMapFunction<String, String>) line -> {
            String[] s = line.split(" ");
            List<String> strings = Arrays.asList(s);
            return strings.iterator();
        });

        JavaPairRDD<Object, Integer> pairWords = words.mapToPair(new PairFunction<String, Object, Integer>() {
            @Override
            public Tuple2<Object, Integer> call(String s) throws Exception {
                return new Tuple2<>(s, 1);
            }
        });

        JavaPairRDD<Object, Integer> result = pairWords.reduceByKey(new Function2<Integer, Integer, Integer>() {
            @Override
            public Integer call(Integer v1, Integer v2) throws Exception {
                return v1 + v2;
            }
        });


        result.foreach(new VoidFunction<Tuple2<Object, Integer>>() {
            @Override
            public void call(Tuple2<Object, Integer> t) throws Exception {
                System.out.println(t);
            }
        });




        sc.stop();
        System.out.println("over");

    }

}

File: scala_learning/TestSpark/src/main/scalacode/cn/wendaocp/scala/spark/WordCount.scala
package cn.wendaocp.scala.spark

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

/**
 * @author liuqiang
 * @since 2020/4/29 11:27
 */
object WordCount {

  def main(args: Array[String]): Unit = {


  }

  def wc(args: Array[String]): Unit = {
    // conf设置
    val conf = new SparkConf()
    conf.setAppName("wordcount")
    conf.setMaster("local")

    // sc 是通往spark集群的唯一通道。 Spark底层操作的就是RDD，只有通过sc才可以创建出RDD
    val sc = new SparkContext(conf)
    sc.setLogLevel("WARN")
    val lines: RDD[String] = sc.textFile("./data/words")

    val words: RDD[String] = lines.flatMap(line => {
      line.split(" ")
    })

    val pairWords: RDD[(String, Int)] = words.map(word => {
      new Tuple2(word, 1)
    }) // (hello, 1) (xxx, 1)

    // 按照key进行分组，每一组的value进行匿名函数指定的方式计算
//    val v: RDD[(String, Int)] = pairWords.reduceByKey((v1: Int, v2: Int) => {
//      v1 + v2
//    })
    val v: RDD[(String, Int)] = pairWords.reduceByKey(_+_) // 因为v1 v2都各自只用了一次，可用 _ 表示

    v.foreach(println)

    println("===")
    // 简化代码
    sc.textFile("./data/words")
      .flatMap(_.split(" "))
      .map((_,1))
      .reduceByKey(_+_)
        .foreach(println)



    sc.stop()

  }


}
